crates/fuel-core/src/coins_query.rs

```diff
@@ -221,6 +274,168 @@ pub async fn random_improve(
+        |coin, _| coin == last_big_coin,
+    )
+    .await
+}
+
+async fn select_coins_until<F>(
+    mut coins_stream: impl Stream<Item = StorageResult<CoinsToSpendIndexEntry>> + Unpin,
+    max: u16,
+    excluded_ids: &ExcludedCoinIds<'_>,
+    predicate: F,
+) -> Result<(u64, Vec<CoinsToSpendIndexEntry>), CoinsQueryError>
+where
+    F: Fn(&CoinsToSpendIndexEntry, u64) -> bool,
+{
+    let mut coins_total_value: u64 = 0;
+    let mut count = 0;
+    let mut coins = Vec::with_capacity(max as usize);
+    while let Some(coin) = coins_stream.next().await {
+        let coin = coin?;
+        if !is_excluded(&coin, excluded_ids)? {
+            if count >= max || predicate(&coin, coins_total_value) {
+                break;
+            }
+            count = count.saturating_add(1);
+            let amount = coin.0.amount();
+            coins_total_value = coins_total_value.saturating_add(amount);
+            coins.push(coin);
+        }
+    }
+    Ok((coins_total_value, coins))
+}
+
+fn is_excluded(
+    (key, coin_type): &CoinsToSpendIndexEntry,
+    excluded_ids: &ExcludedCoinIds,
+) -> Result<bool, CoinsQueryError> {
+    match coin_type {
+        IndexedCoinType::Coin => {
+            let utxo_id_bytes = key.foreign_key_bytes();
+            let tx_id: TxId = utxo_id_bytes
+                .get(..32)
+                .ok_or(CoinsQueryError::IncorrectCoinKeyInIndex)?
+                .try_into()
+                .map_err(|_| CoinsQueryError::IncorrectCoinKeyInIndex)?;
+
+            let output_index = u16::from_be_bytes(
+                utxo_id_bytes
+                    .get(32..34)
+                    .ok_or(CoinsQueryError::IncorrectCoinKeyInIndex)?
+                    .try_into()
+                    .map_err(|_| CoinsQueryError::IncorrectCoinKeyInIndex)?,
+            );
+            Ok(excluded_ids.is_coin_excluded(&UtxoId::new(tx_id, output_index)))
+        }
+        IndexedCoinType::Message => {
+            let nonce_bytes = key.foreign_key_bytes();
+            let nonce: Nonce = nonce_bytes
+                .get(..)
+                .ok_or(CoinsQueryError::IncorrectMessageKeyInIndex)?
+                .try_into()
+                .map_err(|_| CoinsQueryError::IncorrectMessageKeyInIndex)?;
+            Ok(excluded_ids.is_message_excluded(&nonce))
+        }
+    }
+}
+
+fn max_dust_count(max: u16, big_coins_len: u16) -> u16 {
+    let mut rng = rand::thread_rng();
+    rng.gen_range(0..=max.saturating_sub(big_coins_len))
+}
+
+fn skip_big_coins_up_to_amount(
+    big_coins: impl IntoIterator<Item = CoinsToSpendIndexEntry>,
+    mut dust_coins_total: u64,
+) -> impl Iterator<Item = CoinsToSpendIndexEntry> {
+    big_coins.into_iter().skip_while(move |item| {
+        let amount = item.0.amount();
+        dust_coins_total
+            .checked_sub(amount)
+            .map(|new_value| {
+                dust_coins_total = new_value;
+                true
+            })
+            .unwrap_or_default()
```

Comment:
```suggestion
            .unwrap_or(false)
```

Reply:
Updated in 0175e415c2e54fd44c10c0a3df7cc4eacdbf82ce